### PR TITLE
✅ AppcuesViewModel Tests

### PR DESCRIPTION
**Pull request by andretortolano**
Was able to put together some valid tests for AppcuesViewModel

when testing state the idea here is that during setup() we observe and put all states in a list.

then when testing we can check the list size and type check elements

### CODE REVIEW

File: appcues/src/test/java/com/appcues/ui/presentation/AppcuesViewModelTest.kt [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TestRule
+
+@OptIn(ExperimentalCoroutinesApi::class)
+internal class AppcuesViewModelTest {
+
+    @get:Rule
+    val rule: TestRule = InstantTaskExecutorRule()
```
Comment (andretortolano) [inline appcues/src/test/java/com/appcues/ui/presentation/AppcuesViewModelTest.kt:52]: testImplementation 'androidx.arch.core:core-testing:2.2.0'

File: appcues/src/test/java/com/appcues/ui/presentation/AppcuesViewModelTest.kt [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+    private lateinit var viewModel: AppcuesViewModel
+
+    private val uiStates = arrayListOf<UIState>()
+
+    private lateinit var stateJob: Job
+
+    @Before
+    fun setup() {
+        Dispatchers.setMain(UnconfinedTestDispatcher())
```
Comment (andretortolano) [inline appcues/src/test/java/com/appcues/ui/presentation/AppcuesViewModelTest.kt:75]: setting the main dispatcher to UnconfinedTestDispatcher allows us to use viewModelScope from ViewModel() (we do it on init block)

File: appcues/src/main/java/com/appcues/ui/presentation/AppcuesViewModel.kt [hunk 3]
```diff
@@ -21,15 +21,14 @@ import kotlinx.coroutines.flow.MutableStateFlow
...
-import org.koin.core.scope.Scope
 
 internal class AppcuesViewModel(
-    override val scope: Scope,
     private val renderContext: RenderContext,
+    private val coroutineScope: AppcuesCoroutineScope,
+    private val experienceRenderer: ExperienceRenderer,
+    private val actionProcessor: ActionProcessor,
     private val onDismiss: () -> Unit,
-) : ViewModel(), KoinScopeComponent {
```
Comment (andretortolano) [inline appcues/src/main/java/com/appcues/ui/presentation/AppcuesViewModel.kt:32]: changed injection approach to constructor base to simplify test, decouple from the KOIN framework

File: appcues/src/test/java/com/appcues/ui/presentation/AppcuesViewModelTest.kt [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+            every { groupLookup } returns hashMapOf(3 to 0)
+            every { stepIndexLookup } returns hashMapOf(0 to 1234)
+            every { stepContainers } returns listOf(mockStepContainer)
+        }
+        val isFirst = true
+        val mockMetadata: Map<String, Any?> = hashMapOf()
+        // WHEN (flatStepIndex is 2 here)
+        experienceStates.emit(BeginningStep(mockExperience, 0, isFirst, mockMetadata))
+        // THEN
+        assertThat(uiStates).hasSize(1)
```
Comment (iujames) [inline appcues/src/test/java/com/appcues/ui/presentation/AppcuesViewModelTest.kt:137]: for my own understanding, where does uiStates get reset between each test?

**Review by iujames — APPROVED** [review]
great to test all this since the VM is such a core part of rendering, nice job!

File: appcues/src/test/java/com/appcues/ui/presentation/AppcuesViewModelTest.kt (hunk 4 above)
Comment (andretortolano) [reply appcues/src/test/java/com/appcues/ui/presentation/AppcuesViewModelTest.kt:137]: yes, each run the list is a new one

File: appcues/src/test/java/com/appcues/ui/presentation/AppcuesViewModelTest.kt [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+    @Before
+    fun setup() {
+        Dispatchers.setMain(UnconfinedTestDispatcher())
+
+        viewModel = AppcuesViewModel(renderContext, coroutineScope, experienceRenderer, actionProcessor, onDismiss).apply {
+            stateJob = viewModelScope.launch { uiState.collect { uiStates.add(it) } }
+        }
+
+        // clear getStateFlow from history
+        excludeRecords { experienceRenderer.getStateFlow(renderContext) }
```
[the remark this answers is not in the document]
Comment (andretortolano) [reply appcues/src/test/java/com/appcues/ui/presentation/AppcuesViewModelTest.kt:82]: this is excluding the already triggered event from getStateFlow, helps with tests that check for experienceRenderer wasNot called

on each test we have a new experience renderer. before each test run its created an instance of this class and the @Before method is called, followed by the @Test and then the @After, so every run is a new ViewModel nonetheless

**Pull request merged**